iroh-relay/src/client.rs

```diff
@@ -306,356 +164,230 @@ impl ClientBuilder {
+    proxy_url: Option<Url>,
+    key_cache: KeyCache,
+}
 
-    async fn connect_0(&self) -> Result<(Conn, ConnReceiver), ClientError> {
-        let (reader, writer, local_addr) = match self.protocol {
+impl ConnectionBuilder {
+    async fn connect_0(&self) -> Result<(Conn, Option<SocketAddr>)> {
+        let (conn, local_addr) = match self.protocol {
             Protocol::Websocket => {
-                let (reader, writer) = self.connect_ws().await?;
+                let conn = self.connect_ws().await?;
                 let local_addr = None;
-                (reader, writer, local_addr)
+                (conn, local_addr)
             }
             Protocol::Relay => {
-                let (reader, writer, local_addr) = self.connect_derp().await?;
-                (reader, writer, Some(local_addr))
+                let (conn, local_addr) = self.connect_relay().await?;
+                (conn, Some(local_addr))
             }
         };
 
-        let (conn, receiver) =
-            ConnBuilder::new(self.secret_key.clone(), local_addr, reader, writer)
-                .build()
-                .await
-                .map_err(|e| ClientError::Build(e.to_string()))?;
-
-        if self.is_preferred && conn.note_preferred(true).await.is_err() {
-            conn.close().await;
-            return Err(ClientError::Send);
-        }
-
         event!(
             target: "events.net.relay.connected",
             Level::DEBUG,
-            home = self.is_preferred,
             url = %self.url,
+            protocol = ?self.protocol,
         );
 
         trace!("connect_0 done");
-        Ok((conn, receiver))
+        Ok((conn, local_addr))
     }
 
-    async fn connect_ws(&self) -> Result<(ConnReader, ConnWriter), ClientError> {
+    async fn connect_ws(&self) -> Result<Conn> {
         let mut dial_url = (*self.url).clone();
         dial_url.set_path(RELAY_PATH);
         // The relay URL is exchanged with the http(s) scheme in tickets and similar.
         // We need to use the ws:// or wss:// schemes when connecting with websockets, though.
         dial_url
             .set_scheme(if self.use_tls() { "wss" } else { "ws" })
-            .map_err(|()| ClientError::InvalidUrl(self.url.to_string()))?;
+            .map_err(|()| anyhow!("Invalid URL"))?;
 
         debug!(%dial_url, "Dialing relay by websocket");
 
-        let (writer, reader) = tokio_tungstenite_wasm::connect(dial_url).await?.split();
-
-        let cache = self.key_cache.clone();
-
-        let reader = ConnReader::Ws(reader, cache);
-        let writer = ConnWriter::Ws(writer);
-
-        Ok((reader, writer))
+        let conn = tokio_tungstenite_wasm::connect(dial_url).await?;
+        let conn = Conn::new_ws(conn, self.key_cache.clone(), &self.secret_key).await?;
+        Ok(conn)
     }
 
-    async fn connect_derp(&self) -> Result<(ConnReader, ConnWriter, SocketAddr), ClientError> {
+    async fn connect_relay(&self) -> Result<(Conn, SocketAddr)> {
```

Comment:
I've stumbled upon this a bunch.

```suggestion
    async fn connect_derp(&self) -> Result<(Conn, SocketAddr)> {
```

Calling this `connect_relay` as opposed to `connect_ws` is weird. `connect_ws` *also* connects to a relay. And also via the "relay protocol".

The only difference between the two is that `connect_relay` uses the `http derp ...` HTTP upgrade and continues with "derp framing", whereas `_ws` uses `websockets` as HTTP upgrade and uses derp framing.

My opinion is that `_derp` should be kept until the "http derp ..." HTTP upgrade string is removed.

Reply:
Well at that point likely `connect_ws` will be the only version.. naming is weird I agree, not sure which state I prefer at this point

Reply:
Yeah that's okay. I don't hate the name `connect_ws`.